### PR TITLE
ChIP-Seq QC

### PR DESCRIPTION
**Pull request by jvrakor**
@jkokosar @mstajdohar 

## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have their ``re-save`` calls.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of tool should be self explanatory with links to tool's homepage
  and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required: false`` is not needed.
* Add descriptive error and warning messages.
* Place a ``re-checkrc`` call after each process step (call of a tool, a
  Bash command, a Python script or an R script).
* Add ``re-progress`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Changes to Docker images need to be accepted in a separate pull request.
  Before the new images can be used for automated testing of your PR, they
  must be manually tagged and pushed to Docker Hub.
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files with spaces:
  * Genomes: ``genome species.fa.gz``
  * Annotations: ``annotation species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Output and specific input files should include the name of the tested tool.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if length of some list is 0, if string is given where one expects
  integer, if file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

File: resolwe_bio/processes/chip_seq/prepeak_qc.yml [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+        }' "${NAME}.tagAlign" | gzip -c > ${NAME}_tn5.tagAlign.gz
+      {% endif %}
+
+      gzip "${NAME}.tagAlign"
+
+      re-save-file chip_qc "${NAME}_prepeak_qc_report.txt"
+      re-save fraglen $(cut "${NAME}_cc_score_tmp.txt" -f 3)
+      re-save-file tagalign "${NAME}.tagAlign.gz"
+      {% if tn5 %}
+        re-save-file tagalign ${NAME}_tn5.tagAlign.gz
```
Comment (jkokosar) [inline resolwe_bio/processes/chip_seq/prepeak_qc.yml]: Add quotes

File: resolwe_bio/processes/chip_seq/prepeak_qc.yml [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    resources:
+      memory: 16384
+      cores: 10
+  data_name: "{{ alignment.bam.file|basename|default('?') }}"
+  version: 0.0.1
+  type: data:prepeakqc
+  category: ChIP-Seq:QC report
+  flow_collection: sample
+  persistence: CACHED
+  description: >
```
Comment (jkokosar) [inline resolwe_bio/processes/chip_seq/prepeak_qc.yml]: Please add more extensive description, explaining also the output format and results field names. You can use hyperlinks to the web resources explaining key metrics results, e.g. NSC, RSC.

File: resolwe_bio/processes/support_processors/archive.yml [hunk 3]
```diff
@@ -107,18 +121,29 @@
         re-checkrc "Crating IGV session failed"
       done
 
-      expressionmerge_archive.py \
-        --file_path "${FILE_PATH[@]}" \
-        --sample_names "${SAMPLE_NAMES[@]}" \
-        --build "${BUILD[@]}" \
-        --species "${SPECIES[@]}" \
-        --exp_type "${EXP_TYPE[@]}"
+      if [ $(echo -n "${EXP_TYPE[@]}" | wc -c) -ne 0 ]; then
```
Comment (jkokosar) [inline resolwe_bio/processes/support_processors/archive.yml]: Length of the list can be checked by `"${#FILE_PATH[@]}"` as recently accepted to the current master branch

File: resolwe_bio/processes/chip_seq/macs2.yml [hunk 4]
```diff
@@ -32,129 +31,257 @@
...
-            The band width which is used to scan the genome ONLY for model building. You can set this parameter as the sonication fragment size expected from wet experiment. The previous side effect on the peak detection process has been removed. So this parameter only affects the model building.
+            The band width which is used to scan the genome ONLY for model building. You can set
+            this parameter as the sonication fragment size expected from wet experiment. The
+            previous side effect on the peak detection process has been removed. So this parameter
+            only affects the model building.
         - name: nolambda
           label: Use backgroud lambda as local lambda
           type: basic:boolean
           default: false
+          required: false
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/chip_seq/macs2.yml]: The reason for `required:false` on `basic:boolean` inputs is purely aesthetic. I didn't like there is a `*` next to them. Since the bool input's default value is false (frontend) there is no need for it to be required - it makes no sense.

File: resolwe_bio/processes/chip_seq/macs2.yml [hunk 5]
```diff
@@ -231,53 +368,112 @@
...
-      BAND_WIDTH='{% if settings.band_width %} --bw {{settings.band_width}} {% endif %}'
-      DOWN_SAMPLE='{% if settings.down_sample %} --down-sample {% endif %}'
-      BEDGRAPH='{% if settings.bedgraph %} -B {% endif %}'
-      CALL_SUMMITS='{% if settings.call_summits %} --call-summits {% endif %}'
-
-      {% if settings.broad and not settings.call_summits %}
-        BROAD=' --broad {% if settings.broad_cutoff %} --broad-cutoff {{settings.broad_cutoff}}{% endif %}'
+      {% set gsize = species_to_gsize[case.species] %}
+      {% if not gsize %}
+        re-error "{{ case.species }} is not a valid hybrid species. Accepted species are: {{ species_to_gsize.keys() | join(', ') }}"
```
[the remark this answers is not in the document]
Comment (jvrakor) [reply resolwe_bio/processes/chip_seq/macs2.yml]: Good find. I made the same mistake in `macs14` some time ago and it we didn't notice it.
